src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,259 @@
+    pub headers: Option<CredentialSourceHeaders>,
+    pub format: Option<CredentialSourceFormat>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ExternalAccountConfig {
+    pub audience: String,
+    pub subject_token_type: String,
+    pub token_url: String,
+    pub client_id: Option<String>,
+    pub client_secret: Option<String>,
+    pub credential_source: CredentialSource,
+}
+
+#[derive(Debug)]
+struct ExternalAccountTokenProvider {
+    subject_token_provider: Box<dyn SubjectTokenProvider>,
+    config: ExternalAccountConfig,
+}
+
+impl ExternalAccountTokenProvider {
+    pub fn new(config: ExternalAccountConfig) -> Result<ExternalAccountTokenProvider> {
+        let subject_token_provider = subject_token_provider_from_config(config.clone())?;
+        Ok(Self {
+            subject_token_provider,
+            config,
+        })
+    }
+}
+
+/// Detect which subject token provider implementation to use
+fn subject_token_provider_from_config(
+    config: ExternalAccountConfig,
+) -> Result<Box<dyn SubjectTokenProvider>> {
+    if let Some(url) = config.credential_source.url {
+        let creds = UrlSourcedCredentials {
+            url,
+            headers: config.credential_source.headers,
+            format: config.credential_source.format,
+        };
+        return Ok(Box::new(creds));
+    }
+
+    Err(CredentialsError::from_str(
+        false,
+        "unable to parse credential source",
+    ))
+}
+
+#[async_trait::async_trait]
+impl TokenProvider for ExternalAccountTokenProvider {
+    async fn token(&self) -> Result<Token> {
+        let subject_token = self.subject_token_provider.subject_token().await?;
+
+        let audience = self.config.audience.clone();
+        let subject_token_type = self.config.subject_token_type.clone();
+        let url = self.config.token_url.clone();
+        let req = ExchangeTokenRequest {
+            url,
+            audience: Some(audience),
+            subject_token,
+            subject_token_type,
+            authentication: ClientAuthentication {
+                client_id: self.config.client_id.clone(),
+                client_secret: self.config.client_secret.clone(),
+            },
+            ..ExchangeTokenRequest::default()
+        };
+
+        let token_res = STSHandler::exchange_token(req).await?;
+
+        let token = Token {
+            token: token_res.access_token,
+            token_type: token_res.token_type,
+            expires_at: Some(Instant::now() + Duration::from_secs(token_res.expires_in)),
+            metadata: None,
+        };
+        Ok(token)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct ExternalAccountCredentials<T>
+where
+    T: CachedTokenProvider,
+{
+    token_provider: T,
+    quota_project_id: Option<String>,
+}
+
+/// A builder for constructing external account [Credentials] instances.
+///
+/// # Example
+/// ```
+/// # use google_cloud_auth::credentials::external_account::{Builder};
```

Comment:
This example is quite intimidating.  Is there something simpler we can write?  if not, then let's structure the code to be less terrifying (can be in a future PR if this PR is blocking other work).